### ISSUE TITLE
prefer_final_locals should not mark lists later modified with List.add()

### ISSUE THREAD

**Issue by dvorapa** [dart-lang/linter#1849]
**Describe the issue**
Sometimes I get this warning for a list or map declaration:
```List<String> _stale = [];```

**To Reproduce**
```
List<String> _stale = [];
...
if (_rely) _stale.add(_existing);
```

**Expected behavior**
These should never be marked as final per my understanding.

**Additional context**
Latest stable Flutter v1.9.1+h6 with Dart 2.5.0

**Comment by bwilkerson** [comment]
Marking the variable `_stale` as `final` would prevent a new value from being assigned to the variable, but it doesn't prevent the state of the variable's value from being mutated. Adding `final` won't disallow the use of `add`.

**Comment by dvorapa** [comment]
Could you show me some example? I'm not sure, what do you mean.

**Comment by bwilkerson** [comment]
Sure. I'll extend your original example by adding `final` and adding an assignment statement:
```
final List<String> _stale = [];
...
if (_rely) _stale.add(_existing); // No change of semantics from the original, and not an error.
...
if (_clear) _stale = []; // Error because you can't assign to _stale after the initializer.
```

**Comment by dvorapa** [comment]
I see. What about `_stale.clear()` instead of `_stale = []`. This should be fine even when final right?

**Comment by bwilkerson** [comment]
Yes, that would also be fine. The only thing `final` prevents is re-assignment.
